### PR TITLE
Truncate using even length hex string

### PR DESCRIPTION
**Pull request by tcharding**
There are a couple of problems with the width/precision code when we truncate:

- We are printing one too many hex characters if truncation is odd
- The nested precision code (inside width) is doing the wrong comparison

Includes unit tests to verify the behaviour of the new code.

Fix: #70

### CODE REVIEW

**Review by apoelstra — APPROVED** [review]
ACK 2205dc7c0bbc157180ba1f4fc228e4407ef7b2e5

File: src/display.rs [hunk 1]
```diff
@@ -166,8 +165,14 @@ impl<'a> DisplayByteSlice<'a> {
 
         let pad_right = if let Some(width) = f.width() {
             let string_len = match f.precision() {
-                Some(max) if self.bytes.len() * 2 > (max + 1) / 2 => max,
-                Some(_) | None => self.bytes.len() * 2,
+                #[rustfmt::skip]
+                Some(max) => {
+                    // Odd max width which makes no sense for hex strings, honour max but use even length.
```
Comment (tcharding) [inline src/display.rs:170]: Is this statement correct? Yesterday I realised that we may want to support parsing integers in which case `0xa` should be parsable, precision and width should allow `format!("{0:.1}"` in that case, right?

Comment (apoelstra) [reply src/display.rs:170]: Ah, yeah, I think if this is supposed to parse integers (which seems to be the case, given that we're supporting the `0x` prefix) then we ought not to require even lengths.

Comment (tcharding) [reply src/display.rs:170]: Oh boy, the last two hours has show me that parsing hex integers is a whole can of worms. I am starting to see why it was never included in this crate.

File: src/display.rs [hunk 2]
```diff
@@ -166,8 +165,14 @@ impl<'a> DisplayByteSlice<'a> {
...
+                    let max = if max % 2 == 1 { max - 1 } else { max };
```
Comment (Kixunil) [inline src/display.rs:171]: Branchless: `let max = max - (max % 2);` (Maybe the compiler does it I'm not sure.)

**Review by Kixunil — COMMENTED** [review]
Why should we truncate more than what was specified? IMO truncating part of hex digit is valid.

File: src/display.rs (hunk 2 above)
Comment (tcharding) [reply src/display.rs:171]: That suggestion is hot. Legendary.

**Comment by tcharding** [comment]
> Why should we truncate more than what was specified? IMO truncating part of hex digit is valid.

Without doing more invasive changes we cannot support displaying an odd number of hex characters. Therefore it seems more correct to me to print one less than max instead of one more than max, since max is AFAIU an upper limit on how many characters should be displayed.

**Comment by tcharding** [comment]
Used the branchless suggestion and updated the code comment.

**Review by apoelstra — APPROVED** [review]
ACK c7b006902d2c2127fe87327dcd611dcf5281f67f though I agree it would be better if we could support odd-length strings

**Comment by tcharding** [comment]
Changes in force push: `s/ceil/floor/` in code comment, woops.

Also I checked that this does not conflict with #75.

**Review by apoelstra — APPROVED** [review]
ACK 0d7b18ca21da510b1e944427a51ef596082fa221 though I agree it would be better if we could support odd-length strings

**Comment by tcharding** [comment]
Will need rebasing after #77 goes in.

**Comment by tcharding** [comment]
I don't believe this conflicts with #75 but if @Kixunil acks you could merge his first just in case - this one is non-urgent.

**Comment by tcharding** [comment]
Rebased, just for good measure.

**Comment by tcharding** [comment]
Converting to draft because I don't care, I just want a release.

**Pull request closed without merging**